### PR TITLE
feat: Include app permissions with event

### LINKED ISSUE #1014: Include app permissions in events
Events should include App's allowed/disallowed permissions.

Android counterpart: https://github.com/getsentry/sentry-java/issues/1347

### PR DESCRIPTION
**Pull request by kevinrenskers**
## :scroll: Description

We're now including the status of some permissions, namely for Push Notifications, Location access, Media Library and Photo Library. This is added to the app context in every event. These permissions are cached.

We can easily add more permissions, but I figured we can do that in a separate PR, let's get this merged first if we like this approach. Also it would be a lot easier to add more permission once we drop iOS 9 😅

## :bulb: Motivation and Context

Closes #1014

## :green_heart: How did you test it?

With the sample app and unit tests.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by kevinrenskers** [comment]
There is one other problem: at the moment I have hooked my logic up inside of `enrichScope`, to add the new values to the `app` context. But I've notified that this is only done one at SDK start-up (well, technically when `configureScope` is called), not for every event. So I have to undo this change, but what is the correct place to add this info to the event?

File: Sources/Sentry/include/SentryPermissionsObserver.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+typedef enum {
+    kPermissionStatusUnknown,
+    kPermissionStatusGranted,
+    kPermissionStatusDenied
+} SentryPermissionStatus;
+
+@interface SentryPermissionsObserver : NSObject
+
+@property (nonatomic) SentryPermissionStatus hasPushPermission;
```
Comment (kevinrenskers) [inline Sources/Sentry/include/SentryPermissionsObserver.h:14]: A boolean can't be nullable, wrapping it in an NSNumber felt wrong, so I introduced `SentryPermissionStatus` instead.

**Comment by brustolin** [comment]
> Alrighty, the UIApplicationDidBecomeActiveNotification notification is actually called when you grant permissions in the normal happy path flow

That's interesting. If every time a permission popup disappear we receive a `UIApplicationDidBecomeActiveNotification` this might be the solution for almost every permission.

**Comment by kevinrenskers** [comment]
276 test failures in macOS? And I can't run them locally because the test always crashes with `The file “io.sentry” couldn’t be saved in the folder “Caches”`.

**Comment by kevinrenskers** [comment]
I think the UI tests were related to the UIStackView configuration which was breaking things in older iOS versions. Should be fixed now.

**Comment by RomanTysiachnik** [comment]
Hey there,

Could you please share if there'd be a toggle to turn off that feature? 
It is something that may return invalid results or even crash. Both cases are primarily applicable to iOS app extensions.

**Comment by kevinrenskers** [comment]
At the moment we didn't add a toggle for this. Could you share how this could result in a crash?

**Comment by RomanTysiachnik** [comment]
I don't have existing examples, but we've experienced floating crashes inside app extensions, where some features are not expected to be used. Those have been fixed by now, but there might be new cases, so I wish we could have complete control of those requests using our feature toggles on the device.

**Comment by kevinrenskers** [comment]
@philipphofmann @brustolin do you think we should add an option for this? And should it default to on or off?

**Comment by philipphofmann** [comment]
> Hey there,
> 
> Could you please share if there'd be a toggle to turn off that feature? It is something that may return invalid results or even crash. Both cases are primarily applicable to iOS app extensions.

We don't have a toggle for that feature on Android https://github.com/getsentry/sentry-java/pull/2018/, and I think we shouldn't add one if we can safely write the code so it doesn't crash on app extensions. Would that be possible, @kevinrenskers?

**Comment by kevinrenskers** [comment]
Would these permission checks crash in extensions? Is that documented somewhere? I need to look into this. We're just reading permissions, I can hardly believe this would cause a crash 🤔 

I guess I'll create an extension and read these permissions, see what happens. As I've never built an extension before, that'll be a fun exercise by itself :)

**Comment by RomanTysiachnik** [comment]
Yeah, we have an app extension as our primary product on iOS instead of the app itself, so I just want to highlight that 🙃

**Comment by kevinrenskers** [comment]
I've created an IntentUI extension and a Widget extension, and in neither of them did accessing any of these permissions properties cause any problems. Personally I don't think this would ever crash, but just to be sure I added a check to only read permissions when we're running the main app.

**Comment by brustolin** [comment]
> I added a check to only read permissions when we're running the main app

I like this solution, but please create an issue for us to investigate this further in the future, I think this feature could be useful for extensions, especially widgets.

**Review by armcknight — COMMENTED** [review]
This is looking good. I wonder if we couldn't also gather info on the existence of required Plist keys for various permissions requests. The absence of a required perm key in the plist can lead to crashes, so that'd be a quick win diagnostic.

I'm not sure if we can also check for required entitlements in the bundle at runtime, but that's another avenue worth exploring.

See
- https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/requesting_access_to_protected_resources
- https://www.iosdev.recipes/info-plist/permissions/

**Comment by kevinrenskers** [comment]
Asking for permission without the correct things in the info.plist will cause a crash, but merely checking if we have permission doesn't cause a crash (in my testing).

**Comment by armcknight** [comment]
Oh, I wasn't worrying about us causing a crash, more thinking that we could use this as a diagnostic if the app caused such a crash. Do we [plan to] swizzle any CLLocation API?

**Comment by kevinrenskers** [comment]
That might be a nice future feature to add, can you make an issue to keep track of that? I don't think it should necessarily go into this PR. I was really hoping to merge this soon so we can build on top of it.

**Comment by kevinrenskers** [comment]
> > I added a check to only read permissions when we're running the main app
> 
> I like this solution, but please create an issue for us to investigate this further in the future, I think this feature could be useful for extensions, especially widgets.

https://github.com/getsentry/sentry-cocoa/issues/2026

**Comment by kevinrenskers** [comment]
One failure in a benchmarking test. How can I run this locally? And what is the implication? That this new feature is adding overhead?

**Comment by brustolin** [comment]
You can run the test under iOS-Swift sample project, "PerformanceBenchmarks > SentrySDKPerformanceBenchmarkTests.h".
This was added by the profiling team, Im not really sure what it does. 
Can you figure it out whether this is flack or if we should ping then? Thanks!!

**Comment by kevinrenskers** [comment]
Damn, this is the most annoying test to run, ever. It takes forever. The error it runs into on CI is that `usagePercentage` is equal to zero, and it expects it to be higher than zero. This is definitely unrelated to this PR.

**Comment by kevinrenskers** [comment]
Pushed another commit without changes and yea now they succeed just fine. I think the profiling team needs to look at their tests.

**Review by brustolin — APPROVED** [review]
LGTM

**Pull request merged**